plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+#include "wayfire/core.hpp"
+
+#include "wayfire/seat.hpp"
+#include "wayfire/scene-input.hpp"
+#include "wayfire/scene-operations.hpp"
+#include "wayfire/unstable/wlr-surface-node.hpp"
+#include "wayfire/unstable/wlr-surface-controller.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/output-layout.hpp"
+#include "wayfire/signal-definitions.hpp"
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/util.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/plugins/common/simple-text-node.hpp>
+#include <wayfire/util/log.hpp>
+
+enum lock_state
+{
+    LOCKING,
+    LOCKED,
+    UNLOCKED,
+    DESTROYED,
+    ZOMBIE,
+} state;
+
+class wf_session_lock_plugin : public wf::plugin_interface_t
+{
+    class wayfire_session_lock;
+
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
+    {
+      public:
+        lock_surface_node_t(wayfire_session_lock *lock,
+            wlr_session_lock_surface_v1 *lock_surface,
+            wf::output_t *output) :
+            wf::scene::wlr_surface_node_t(lock_surface->surface, true /* autocommit */),
+            lock(lock),
+            lock_surface(lock_surface),
+            output(output),
+            interaction(std::make_unique<lock_surface_keyboard_interaction>(lock_surface->surface,
+                lock_surface->output))
+        {
+            lock_surface_destroy.set_callback([this] (void *data)
+            {
+                wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+                wf::scene::remove_child(shared_from_this());
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
+                const char *name = this->output->handle ? this->output->handle->name : "(deleted)";
+                LOGD("lock_surface on ", name, " destroyed");
+            });
+            lock_surface_destroy.connect(&lock_surface->events.destroy);
+        }
+
+        void attach_to_layer()
+        {
+            auto layer_node = output->node_for_layer(wf::scene::layer::LOCK);
+            wf::scene::add_front(layer_node, shared_from_this());
+            wf::wlr_surface_controller_t::create_controller(lock_surface->surface, layer_node);
+            focus();
+        }
+
+        void focus()
+        {
+            wf::get_core().seat->set_active_node(shared_from_this());
+            interaction->handle_keyboard_enter(wf::get_core().seat.get());
```

Comment:
Why is this necessary? A simple `refocus()` call should be enough to focus the node. Manually calling `enter()` is usually not what you want, this is what core uses when it focuses your node.